integrations/azure-devops/azure_devops/client.py

```diff
@@ -0,0 +1,268 @@
+        params = {"includeCapabilities": "true"}
+        projects_url = f"{self._organization_base_url}/{API_URL_PREFIX}/projects"
+        async for projects in self._get_paginated_by_top_and_continuation_token(
+            projects_url, additional_params=params
+        ):
+            yield projects
+
+    @cache_results("teams")
+    async def generate_teams(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        teams_url = f"{self._organization_base_url}/{API_URL_PREFIX}/teams"
+        async for teams in self._get_paginated_by_top_and_skip(teams_url):
+            yield teams
+
+    async def generate_members(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for teams in self.generate_teams():
+            for team in teams:
+                members_in_teams_url = f"{self._organization_base_url}/{API_URL_PREFIX}/projects/{team['projectId']}/teams/{team['id']}/members"
+                async for members in self._get_paginated_by_top_and_skip(
+                    members_in_teams_url
+                ):
+                    for member in members:
+                        member["teamId"] = team["id"]
+                    yield members
+
+    @cache_results("repositories")
+    async def generate_repositories(self) -> AsyncGenerator[list[dict[Any, Any]], None]:
+        async for projects in self.generate_projects():
+            for project in projects:
+                repos_url = f"{self._organization_base_url}/{project['id']}/{API_URL_PREFIX}/git/repositories"
+                repositories = self._parse_response_values(
+                    await self.send_request("GET", repos_url)
+                )
+                yield repositories
+
+    async def generate_pull_requests(
+        self, search_filters: Optional[dict[str, Any]] = None
+    ) -> AsyncGenerator[list[dict[Any, Any]], None]:
+        async for repositories in self.generate_repositories():
+            for repository in repositories:
+                pull_requests_url = f"{self._organization_base_url}/{repository['project']['id']}/{API_URL_PREFIX}/git/repositories/{repository['id']}/pullrequests"
+                async for filtered_pull_requests in self._get_paginated_by_top_and_skip(
+                    pull_requests_url, search_filters
+                ):
+                    yield filtered_pull_requests
+
+    @cache_results("pipelines")
+    async def generate_pipelines(self) -> AsyncGenerator[list[dict[Any, Any]], None]:
+        async for projects in self.generate_projects():
+            for project in projects:
+                pipelines_url = f"{self._organization_base_url}/{project['id']}/{API_URL_PREFIX}/pipelines"
+                async for (
+                    pipelines
+                ) in self._get_paginated_by_top_and_continuation_token(pipelines_url):
+                    for pipeline in pipelines:
+                        pipeline["projectId"] = project["id"]
```

Comment:
i dont like setting values into the give data
pipeline really does not have project id?

Also the convention we made is if we really decided to set a custom value then we should prefix it with double `_`

Reply:
A repository Id can be fetched, but the way to actually get this is not pretty
Let's talk about this